src/Enhavo/Bundle/UserBundle/EventListener/UserAuthenticationSubscriber.php

```diff
@@ -0,0 +1,139 @@
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class UserAuthenticationSubscriber implements EventSubscriberInterface
+{
+    public function __construct(
+        private UserManager           $userManager,
+        private ConfigurationProvider $configurationProvider,
+        private RequestStack          $requestStack,
+        private RouterInterface       $router,
+        private TranslatorInterface   $translator,
+    )
+    {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            UserEvent::class => 'onUserEvent',
+            UserLoginEvent::class => 'onLogin',
+            UserLoginFailureEvent::class => 'onLoginFailure',
+        ];
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function onLogin(UserLoginEvent $event): void
+    {
+        $user = $event->getUser();
+        $configKey = $this->getConfigKey();
+        $configuration = $this->configurationProvider->getLoginConfiguration($configKey);
+        $updated = $user->getPasswordUpdatedAt();
+
+        $user->setLastLogin(new \DateTime());
+        $user->setFailedLoginAttempts(0);
+        $user->setLastFailedLoginAttempt(null);
+
+        if ($configuration->getPasswordMaxAge() &&
+            (!$updated || $updated->modify(sprintf('+%s', $configuration->getPasswordMaxAge())) < new \DateTime())
+        ) {
+            // todo: This can be handled by PasswordMaxAgeResetHandler
+            $this->userManager->logout();
+
+            $user->setPlainPassword(random_bytes(12));
+
+            $configuration = $this->configurationProvider->getResetPasswordRequestConfiguration($configKey);
+            $this->userManager->resetPassword($user, $configuration);
+            $url = $this->generateUrl($configuration->getRedirectRoute());
+
+            $this->addError($this->getRequest()->getSession(), 'login.error.password_expired');
+            $event->setResponse(new RedirectResponse($url));
+
+        } else {
+            $this->userManager->update($user);
+        }
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function onLoginFailure(UserLoginFailureEvent $event): void
+    {
+        $user = $event->getUser();
+
+        if ($user) {
+            $configKey = $this->getConfigKey();
+            $configuration = $this->configurationProvider->getLoginConfiguration($configKey);
+            if ($configuration->getMaxFailedLoginAttempts() > 0) {
+
+                $user->setFailedLoginAttempts(1 + $user->getFailedLoginAttempts());
+                $user->setLastFailedLoginAttempt(new \DateTime());
+
+                if ($user->getFailedLoginAttempts() > $configuration->getMaxFailedLoginAttempts()) {
+                    // todo: this can be handled by LoginFailureResetHandler
+                    $user->setPlainPassword(random_bytes(12));
+                    $user->setFailedLoginAttempts(0);
```

Comment:
Also here, the user should see the fail messages again. After he logged in successfully we clear the login attempts state